third_party/move/move-vm/integration-tests/src/tests/native_tests.rs

```diff
@@ -29,65 +32,122 @@ fn make_failed_native() -> NativeFunction {
-            ident_str!("foo2"),
-            &[],
-            vec![],
-        )
-        .unwrap_err();
-        assert_eq!(err.exec_state().unwrap().stack_trace().len(), 1);
+        }}
+        "#,
+        TEST_ADDR.to_hex(),
+    );
+    compile_and_publish(&mut storage, code);
+    let module_storage = storage.as_unsync_code_storage();
+
+    let err = execute_function_for_test(
+        &storage,
+        &module_storage,
+        &module_id,
+        ident_str!("foo"),
+        &[],
+        vec![],
+    )
+    .unwrap_err();
+    assert!(err.exec_state().unwrap().stack_trace().is_empty());
+
+    let err = execute_function_for_test(
+        &storage,
+        &module_storage,
+        &module_id,
+        ident_str!("foo2"),
+        &[],
+        vec![],
+    )
+    .unwrap_err();
+    assert_eq!(err.exec_state().unwrap().stack_trace().len(), 1);
+}
+
+fn make_load_module_b() -> NativeFunction {
+    Arc::new(move |_, _, _| -> PartialVMResult<NativeResult> {
+        Ok(NativeResult::LoadModule {
+            module_name: ModuleId::new(TEST_ADDR, ident_str!("b").to_owned()),
+        })
+    })
+}
+
+#[test_case(true)]
+#[test_case(false)]
+fn test_load_module_native_result(enable_lazy_loading: bool) {
+    let a_id = ModuleId::new(TEST_ADDR, ident_str!("a").to_owned());
+    let natives = vec![(
+        TEST_ADDR,
+        ident_str!("a").to_owned(),
+        ident_str!("load_module_b").to_owned(),
+        make_load_module_b(),
+    )];
+    let runtime_environment = RuntimeEnvironment::new_with_config(natives, VMConfig {
+        enable_lazy_loading,
+        ..VMConfig::default()
+    });
+    let mut storage = InMemoryStorage::new_with_runtime_environment(runtime_environment);
+
+    let code_a = format!(
+        r#"
+        module 0x{0}::a {{
+            fun foo() {{ load_module_b(); }}
+            native fun load_module_b();
+        }}
+        "#,
+        TEST_ADDR.to_hex(),
+    );
+    compile_and_publish(&mut storage, code_a);
+
+    let mut add_module = |m: CompiledModule| {
+        let mut blob = vec![];
+        m.serialize(&mut blob).unwrap();
+        storage.add_module_bytes(m.self_addr(), m.self_name(), blob.into());
+        m.self_id()
+    };
+
+    let b = empty_module_with_dependencies_and_friends_at_addr(TEST_ADDR, "b", vec!["c"], vec![]);
+    let c = empty_module_with_dependencies_and_friends_at_addr(TEST_ADDR, "c", vec!["d"], vec![]);
+    let d = empty_module_with_dependencies_and_friends_at_addr(TEST_ADDR, "d", vec![], vec![]);
+
+    let b_id = add_module(b);
+    let c_id = add_module(c);
+    let d_id = add_module(d);
+
+    let code_storage = storage.as_unsync_code_storage();
+    assert_ok!(execute_function_for_test(
+        &storage,
+        &code_storage,
+        &a_id,
+        ident_str!("foo"),
+        &[],
+        vec![],
+    ));
+
+    if enable_lazy_loading {
+        code_storage
+            .module_storage()
+            .assert_cached_state(vec![&b_id], vec![&a_id]);
```

Comment:
what does assert_cached_state do? in general a couple one liners to explain the testing logic would help (the dependency graph is pretty clear, just how execute_function_for_test is supposed to cause the divergence here)

Reply:
Indeed!

Asserting cached state checks if modules are in verified or deserialized states, and if they are cached. There is a docstring for the method, but I agree that explaining the test here is better, adding comments👍